core/src/verify.rs

```diff
@@ -199,52 +213,36 @@ impl CommitSequenceVerifier {
         if rs.genesis_info != self.reserved_state.genesis_info {
             return Err(Error::InvalidArgument("genesis_info changes".to_string()));
         }
-        // Check that the newly added (if exists) `Member::name` is unique.
-        let existing_names = self
-            .reserved_state
-            .members
-            .iter()
-            .map(|m| &m.name)
-            .collect::<HashSet<_>>();
+        // Check that `Member::name` and `Member::public_key` are unique.
         for member in &rs.members {
-            if !existing_names.contains(&member.name)
-                && rs.members.iter().filter(|m| m.name == member.name).count() > 1
-            {
+            if rs.members.iter().filter(|m| m.name == member.name).count() > 1 {
                 return Err(Error::InvalidArgument(format!(
                     "member name '{}' already exists",
                     member.name
                 )));
             }
-        }
-        // Check that `member` monotonically increases (refer to `Member::expelled`).
-        let existing_members = self
-            .reserved_state
-            .members
-            .iter()
-            .map(|m| m.public_key.clone())
-            .collect::<HashSet<_>>();
-        for member in &rs.members {
-            if existing_members.contains(&member.public_key) && member.expelled {
+            if rs
+                .members
+                .iter()
+                .filter(|m| m.public_key == member.public_key)
+                .count()
+                > 1
```

Comment:
I think this approach is fine and valid but would be better to introduce `HashSet` structure given the cases where the size of members goes large. 

The suggestion code block shown below utilizes `HashMap::insert` which is used to check duplicates. The method returns `false` if the value was already assigned in the set before.

```suggestion
// Check that `Member::name` and `Member::public_key` are unique.
let mut member_names = HashSet::new();
let mut public_keys = HashSet::new();

for member in &rs.members {
    if !member_names.insert(&member.name) {
        return Err(Error::InvalidArgument(format!(
            "member name '{}' already exists",
            member.name
        )));
    }

    if !public_keys.insert(&member.public_key) {
        return Err(Error::InvalidArgument(format!(
            "the public key of '{}' already exists",
            member.name
        )));
    }
}
```